core-bundle/src/EventListener/DataContainer/Undo/LabelListener.php

```diff
@@ -0,0 +1,195 @@
+                if ('date' === $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['eval']['rgxp']) {
+                    $labelValues[$k] = $arrRow[$v] ? Date::parse(Config::get('dateFormat'), $arrRow[$v]) : '-';
+                } elseif ('time' === $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['eval']['rgxp']) {
+                    $labelValues[$k] = $arrRow[$v] ? Date::parse(Config::get('timeFormat'), $arrRow[$v]) : '-';
+                } else {
+                    $labelValues[$k] = $arrRow[$v] ? Date::parse(Config::get('datimFormat'), $arrRow[$v]) : '-';
+                }
+                continue;
+            }
+
+            if (($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['eval']['isBoolean'] ?? null) || ('checkbox' === $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['inputType'] && !$GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['eval']['multiple'])) {
+                $labelValues[$k] = $arrRow[$v] ? $this->translator->trans('MSC.yes', [], 'contao_default') : $this->translator->trans('MSC.no', [], 'contao_default');
+                continue;
+            }
+
+            $row_v = StringUtil::deserialize($arrRow[$v]);
+
+            if (\is_array($row_v)) {
+                $args_k = [];
+
+                foreach ($row_v as $option) {
+                    $args_k[] = $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$option] ?: $option;
+                }
+
+                $labelValues[$k] = implode(', ', $args_k);
+                continue;
+            }
+
+            if (isset($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]])) {
+                $labelValues[$k] = \is_array($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]]) ? $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]][0] : $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['reference'][$arrRow[$v]];
+                continue;
+            }
+
+            if (($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['eval']['isAssociative'] ?? null) || (ArrayUtil::isAssoc($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['options'] ?? null) && isset($GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['options'][$arrRow[$v]]))) {
+                $labelValues[$k] = $GLOBALS['TL_DCA'][$dc->table]['fields'][$v]['options'][$arrRow[$v]];
+                continue;
+            }
+
+            $labelValues[$k] = $arrRow[$v];
+        }
+
+        $label = $labelConfig['format'] ? vsprintf($labelConfig['format'], $labelValues) : implode(', ', $labelValues);
+
+        // Remove empty brackets (), [], {}, <> and empty tags from the label
+        $label = preg_replace('/\( *\) ?|\[ *] ?|{ *} ?|< *> ?/', '', $label);
+        $label = preg_replace('/<[^>]+>\s*<\/[^>]+>/', '', $label);
```

Comment:
What is this doing?

Reply:
See https://github.com/contao/contao/pull/3644 - it's part of the label creation in the DataContainer.

Reply:
If we need to duplicate this magic, should we rather put it into `StringUtil` and make it available for anyone?

Reply:
If #3644 gets merged `DataContainer::generateRecordLabel` can be used in this PR as well an nothing needs to be duplicated.

Reply:
I now use the  `generateRecordLabel` method 👍